File: packages/analysis/src/generate-deletes.ts
/** Return every unique Unicode code-point deletion through `maxEdits`. */
export function generateDeletes(term: string, maxEdits: 1 | 2): string[] {
  let frontier = new Set([term]);
  const all = new Set(frontier);
  for (let depth = 0; depth < maxEdits; depth++) {
    const next = new Set<string>();
    for (const variant of frontier) {
      const chars = [...variant];
      for (let index = 0; index < chars.length; index++) {
        next.add(
          chars.slice(0, index).join("") + chars.slice(index + 1).join(""),
        );
      }
    }
    for (const variant of next) all.add(variant);
    frontier = next;
  }
  return [...all];
}

File: python/searchable-analysis/src/searchable_analysis/generate_deletes.py
def generate_deletes(term: str, max_edits: int) -> list[str]:
    frontier = {term}
    all_variants = set(frontier)
    for _ in range(max_edits):
        next_frontier: set[str] = set()
        for variant in frontier:
            chars = list(variant)
            for index in range(len(chars)):
                next_frontier.add("".join(chars[:index] + chars[index + 1 :]))
        all_variants.update(next_frontier)
        frontier = next_frontier
    return list(all_variants)
